### ISSUE TITLE
index(self, indices) only dispatches on self and ignores DispatchKeys in indices

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#60461]
## 🐛 Bug

`func: index.Tensor(Tensor self, Tensor?[] indices) -> Tensor` seems to be ignoring the dispatch keys in `indices` and only dispatches according to `self`.

## To Reproduce

```
import torch
num_elements=10
r = torch.ones(num_elements, dtype=torch.float, device='cpu')
scale = 1.0
zero_point = 2
qr = torch.quantize_per_tensor(r, scale, zero_point, torch.qint8)
x = torch.randn(10)

# x is strided CPUTensor, qr is QuantizedCPU tensor
x[qr]
```

GDB-ing through this shows that the dispatcher calls native::index instead of native::quantized_index.

## Expected behavior
 
Since QuantizedCPU has higher priority than CPU the above should have dispatched to the QuantizedCPU key.

## Environment
master

## Additional context

The context for this is actually to support functorch: https://github.com/facebookresearch/functorch/issues/59

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo

**Comment by zou3519** [comment]
This sounds very reminiscent of https://github.com/pytorch/pytorch/pull/58283, cc @Chillee @bhosmer

**Comment by ngimel** [comment]
Please note that actual quantized_index implementation assumes only `self` is quantized https://github.com/pytorch/pytorch/blob/0126f4284129966898447d66dda4fb237cdb2585/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L379-L398, so it's a problem of lax error checking for indices type (similar to #59140) rather than dispatch

**Comment by zou3519** [comment]
[The backtrace](https://gist.github.com/zou3519/6f02eddb6b397a506f156434648b8771) shows that VariableType::index_Tensor dispatches to native::index. Because native_functions.yaml specifies dispatch for QuantizedCPU, I would have expected the dispatcher to dispatch instead to native::quantized_index; native::quantized_index should be responsible for raising an error since it assumes only `self` is quantized.

There is a pretty suspicious comment over at https://github.com/pytorch/pytorch/blob/71b83c27e2d5210546473212322090234a7cced0/aten/src/ATen/native/native_functions.yaml#L2190 though...

**Comment by zou3519** [comment]
Removed quantization label because this isn't only a quantization problem; I just used quantization to demonstrate the problem in the dispatcher.
